File: backend/app/schemas/news.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class NewsNormalizeCreate(BaseModel):
    workspace_code: str
    source_types: list[str] = Field(default_factory=list)
    limit: int | None = Field(default=None, ge=0)


class NewsNormalizeRead(BaseModel):
    workspace_code: str
    raw_scanned: int
    news_created: int
    news_updated: int
    raw_skipped: int
    dedupe_groups_updated: int
    winners: int
    losers: int


class NewsItemRead(BaseModel):
    id: str
    workspace_code: str
    domain_code: str
    raw_item_id: str
    data_source_id: str
    source_type: str
    source_name: str
    source_url: str | None
    canonical_url: str | None
    source_title: str
    normalized_title: str
    summary: str
    author: str
    published_at: datetime | None
    focus_id: int
    dedupe_key: str
    active: bool
    duplicate_of_id: str | None
    normalization_status: str
    normalization_notes: str


class DedupeGroupItemRead(BaseModel):
    id: str
    news_item_id: str
    is_winner: bool
    duplicate_reason: str
    rank_score: float
    title: str
    source_type: str
    source_name: str
    source_url: str | None


class DedupeGroupRecommendationRead(BaseModel):
    run_id: str
    run_key: str
    day_key: str | None
    recommendation_item_id: str
    rank: int
    selected: bool
    final_score: float
    quality_score: float
    topic_score: float
    freshness_score: float
    feedback_score: float
    diversity_score: float
    source_score: float
    heat_score: float
    recommendation_reason: str
    admission_level: str
    admission_score: float
    admission_pool: str
    noise_types: list[str] = Field(default_factory=list)
    reject_reasons: list[str] = Field(default_factory=list)
    scorer_breakdown: dict[str, object] = Field(default_factory=dict)
    expert_routes: list[str] = Field(default_factory=list)


class DedupeGroupDailyReportRead(BaseModel):
    daily_report_id: str
    daily_report_item_id: str
    day_key: str
    report_status: str
    adoption_status: int
    generated_news_id: str
    generation_status: str
    category: str


class DedupeGroupLineageNodeRead(BaseModel):
    object_type: str
    object_id: str | None
    label: str
    status: str
    review_note: str
    target_path: str | None = None
    occurred_at: datetime | None = None
    metadata: dict[str, object] = Field(default_factory=dict)


class DedupeGroupLineageRead(BaseModel):
    nodes: list[DedupeGroupLineageNodeRead] = Field(default_factory=list)


class DedupeGroupRead(BaseModel):
    id: str
    workspace_code: str
    domain_code: str
    dedupe_key: str
    winner_news_item_id: str | None
    winner_title: str | None
    winner_published_at: datetime | None = None
    winner_source_type: str | None = None
    item_count: int
    status: str
    items: list[DedupeGroupItemRead] = Field(default_factory=list)
    recommendation: DedupeGroupRecommendationRead | None = None
    daily_report: DedupeGroupDailyReportRead | None = None
    lineage: DedupeGroupLineageRead = Field(default_factory=DedupeGroupLineageRead)

File: frontend/src/api/news.ts
import { requestJson } from "./http";

export interface NewsItemRecord {
  id: string;
  workspace_code: string;
  domain_code: string;
  raw_item_id: string;
  data_source_id: string;
  source_type: string;
  source_name: string;
  source_url: string | null;
  canonical_url: string | null;
  source_title: string;
  normalized_title: string;
  summary: string;
  author: string;
  published_at: string | null;
  focus_id: number;
  dedupe_key: string;
  active: boolean;
  duplicate_of_id: string | null;
  normalization_status: string;
  normalization_notes: string;
}

export interface DedupeGroupItemRecord {
  id: string;
  news_item_id: string;
  is_winner: boolean;
  duplicate_reason: string;
  rank_score: number;
  title: string;
  source_type: string;
  source_name: string;
  source_url: string | null;
}

export interface DedupeGroupRecommendationRecord {
  run_id: string;
  run_key: string;
  day_key: string | null;
  recommendation_item_id: string;
  rank: number;
  selected: boolean;
  final_score: number;
  quality_score: number;
  topic_score: number;
  freshness_score: number;
  feedback_score: number;
  diversity_score: number;
  source_score: number;
  heat_score: number;
  recommendation_reason: string;
  admission_level: string;
  admission_score: number;
  admission_pool: string;
  noise_types: string[];
  reject_reasons: string[];
  scorer_breakdown: Record<string, unknown>;
  expert_routes: string[];
}

export interface DedupeGroupDailyReportRecord {
  daily_report_id: string;
  daily_report_item_id: string;
  day_key: string;
  report_status: string;
  adoption_status: number;
  generated_news_id: string;
  generation_status: string;
  category: string;
}

export interface DedupeGroupLineageNodeRecord {
  object_type: string;
  object_id: string | null;
  label: string;
  status: string;
  review_note: string;
  target_path: string | null;
  occurred_at: string | null;
  metadata: Record<string, unknown>;
}

export interface DedupeGroupLineageRecord {
  nodes: DedupeGroupLineageNodeRecord[];
}

export interface DedupeGroupRecord {
  id: string;
  workspace_code: string;
  domain_code: string;
  dedupe_key: string;
  winner_news_item_id: string | null;
  winner_title: string | null;
  winner_published_at: string | null;
  winner_source_type: string | null;
  item_count: number;
  status: string;
  items: DedupeGroupItemRecord[];
  recommendation: DedupeGroupRecommendationRecord | null;
  daily_report: DedupeGroupDailyReportRecord | null;
  lineage: DedupeGroupLineageRecord;
}

export interface NewsNormalizeResult {
  workspace_code: string;
  raw_scanned: number;
  news_created: number;
  news_updated: number;
  raw_skipped: number;
  dedupe_groups_updated: number;
  winners: number;
  losers: number;
}

export async function fetchDedupeGroups(
  workspaceCode: string,
  limit = 80,
  filters: {
    q?: string;
    recommendationStatus?: string;
    dailyStatus?: string;
    admissionLevel?: string;
    sourceType?: string;
    sort?: string;
  } = {}
): Promise<DedupeGroupRecord[]> {
  const params = new URLSearchParams({ workspace_code: workspaceCode, limit: String(limit) });
  if (filters.q) {
    params.set("q", filters.q);
  }
  if (filters.recommendationStatus && filters.recommendationStatus !== "all") {
    params.set("recommendation_status", filters.recommendationStatus);
  }
  if (filters.dailyStatus && filters.dailyStatus !== "all") {
    params.set("daily_status", filters.dailyStatus);
  }
  if (filters.admissionLevel) {
    params.set("admission_level", filters.admissionLevel);
  }
  if (filters.sourceType) {
    params.set("source_type", filters.sourceType);
  }
  if (filters.sort) {
    params.set("sort", filters.sort);
  }
  return requestJson<DedupeGroupRecord[]>(`/api/dedupe-groups?${params.toString()}`);
}

export async function fetchNewsItems(
  workspaceCode: string,
  active: boolean | null = null,
  limit = 80
): Promise<NewsItemRecord[]> {
  const params = new URLSearchParams({ workspace_code: workspaceCode, limit: String(limit) });
  if (active !== null) {
    params.set("active", String(active));
  }
  return requestJson<NewsItemRecord[]>(`/api/news-items?${params.toString()}`);
}

export async function normalizeNewsItems(
  workspaceCode: string,
  sourceTypes: string[] = [],
  limit: number | null = null
): Promise<NewsNormalizeResult> {
  return requestJson<NewsNormalizeResult>("/api/news-items/normalize", {
    method: "POST",
    body: JSON.stringify({
      workspace_code: workspaceCode,
      source_types: sourceTypes,
      limit
    })
  });
}
